pkg/oracle/neofs/neofs.go

```diff
@@ -0,0 +1,174 @@
+package neofs
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/mr-tron/base58"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neofs-api-go/pkg/client"
+	"github.com/nspcc-dev/neofs-api-go/pkg/container"
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+	objectv2 "github.com/nspcc-dev/neofs-api-go/v2/object"
+)
+
+const (
+	// URIScheme is the name of neofs URI scheme.
+	URIScheme = "neofs"
+
+	// containerIDSize is the size of container id in bytes.
+	containerIDSize = sha256.Size
+	// objectIDSize is the size of container id in bytes.
+	objectIDSize = sha256.Size
+	// rangeSep is a separator between offset and length.
+	rangeSep = '|'
+
+	rangeCmd  = "range"
+	headerCmd = "header"
+	hashCmd   = "hash"
+)
+
+// Various validation errors.
+var (
+	ErrInvalidScheme    = errors.New("invalid URI scheme")
+	ErrMissingObject    = errors.New("object ID is missing from URI")
+	ErrInvalidContainer = errors.New("container ID is invalid")
+	ErrInvalidObject    = errors.New("object ID is invalid")
+	ErrInvalidRange     = errors.New("object range is invalid (expected 'Offset|Length'")
+	ErrInvalidCommand   = errors.New("invalid command")
+)
+
+// Get returns neofs object from the provided url.
+// URI scheme is "neofs://<Container-ID>/<Object-ID/<Command>/<Params>".
+// If Command is not provided, full object is requested.
+// TODO multiple addresses
+// 	Addresses are queried sequentially using the same context.
+func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string) ([]byte, error) {
+	if u.Scheme != URIScheme {
+		return nil, ErrInvalidScheme
+	}
+	ps := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
+	if len(ps) == 0 {
+		return nil, ErrMissingObject
+	}
+
+	rawCID, err := base58.Decode(u.Hostname())
+	if err != nil || len(rawCID) != containerIDSize {
+		return nil, ErrInvalidContainer
+	}
+	rawOID, err := base58.Decode(ps[0])
+	if err != nil || len(rawOID) != objectIDSize {
+		return nil, ErrInvalidObject
+	}
+
+	var cid, oid [32]byte
+	copy(cid[:], rawCID)
+	containerID := container.NewID()
+	containerID.SetSHA256(cid)
+
+	copy(oid[:], rawOID)
+	objectID := object.NewID()
+	objectID.SetSHA256(oid)
+
+	objectAddr := object.NewAddress()
+	objectAddr.SetContainerID(containerID)
+	objectAddr.SetObjectID(objectID)
```

Comment:
With latest neofs-api-go you can simplify parsing a bit:

```go
	containerID := container.NewID()
	err = containerID.Parse(u.Hostname())
 	if err != nil {
		return nil, ErrInvalidContainer
	}

	objectID := object.NewID()
	err = objectID.Parse(ps[0])
 	if err != nil {
		 return nil, ErrInvalidObject
	}

	objectAddr := object.NewAddress()
	objectAddr.SetContainerID(containerID)
	objectAddr.SetObjectID(objectID)
```

Reply:
Still relevant: we can replace base58 encoding and copying SHA256 data with `containerID.Parse(u.Hostname())` for container and  `objectID.Parse(ps[0])` for object.

Reply:
@fyrchik?

Reply:
Fixed.